src/NBitcoin/ConsensusChainIndexer.cs

```diff
@@ -0,0 +1,215 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace NBitcoin
+{
+    public class ConsensusChainIndexer
+    {
+        private readonly object lockObject = new object();
+        private readonly Dictionary<int, ChainedHeader> blocksByHeight = new Dictionary<int, ChainedHeader>();
+        private readonly Dictionary<uint256, ChainedHeader> blocksById = new Dictionary<uint256, ChainedHeader>();
+
+        public Network Network { get; }
+        public ChainedHeader Tip { get; private set; }
+
+        public int Height => this.Tip.Height;
+        public ChainedHeader Genesis => this.GetBlock(0);
+
+        public ConsensusChainIndexer(Network network)
+        {
+            this.Network = network;
+
+            this.Initialize(new ChainedHeader(network.GetGenesis().Header, network.GetGenesis().GetHash(), 0));
+        }
+
+        public ConsensusChainIndexer(Network network, ChainedHeader chainedHeader) 
+        {
+            this.Network = network;
+
+            this.Initialize(chainedHeader);
+        }
+
+        public void Initialize(ChainedHeader chainedHeader)
+        {
+            this.blocksById.Clear();
```

Comment:
I think this should be inside the `lock. Even if we don't expect this to be used concurrently before being initialized it's happened before.

Reply:
Totally agree I will fix it.